### DISCUSSION TITLE
Testing

### DISCUSSION THREAD

**Discussion by alvarosabu** [Tresjs/tres#231] [Ideas]
During the development of the library, I found having a decent level of test coverage was quite complicated for 2 main reasons 🤦:

- Abstract Vue & ThreeJS functionality and test only **our** code 
- Faking canvas and dom elements on node environment (vitest) is tricky 

A week ago I worked with @userquin on testing Vitest Browser Runner with TresJS and the results were amazing.

![FuBPgUaXsAcjoaA](https://user-images.githubusercontent.com/4699008/233982594-29d44962-1fc0-49ac-82eb-ef6dbfb282c4.jpg)

This would solve the second constraint listed above and open the door to be able to mount the components and test them 👏🏻.

I think is important we provide a decent level of testing on the ecosystem, starting with `core` and `cientos` packages, especially now that we are close to opening to the public

- Composables and utils should be covered easily
- Testing on Vue components such as abstractions (`cientos`) will not focus on DOM (what is normally done with  https://test-utils.vuejs.org/ but rather test the scene state)
- There is no sense in testing a canvas. We can only test the state.

Example:

Testing [`Box.vue`](https://github.com/Tresjs/cientos/blob/main/src/core/Box.vue)

```
describe('Box', () => {
  it('should work', () => {
    const wrapper = mount(TheExperience, {
      attachTo: document.body,
      components: {
        Box,
      },
    })
    // Get state somehow
     const box = wrapper.vm.state.scene.find('TresBoxGeometry')
    expect(box).toExist()
  })
})
```

First, let's define what types of tests we want. @Tresjs/cientos @Tresjs/core 

## Follow up tasks

- [ ] Core
- [ ] Cientos

**Comment by FreakTheMighty** [comment]
@alvarosabu was there any progress made on this? I've been trying to instrument integration tests and have really struggled to get past mocking webgl.
